### PR TITLE
ReactOS boot support bringup PR (3 of 3)

### LINKED ISSUE #17: Add ReactOS boot support
Since the ReactOS port to Xbox has been recently revived, it's likely that users may want to boot ReactOS in addition to Linux.

@binarymaster has added this support recently [here](https://github.com/binarymaster/cromwell/tree/master), and those patches should be ported over to this repository.

### PR DESCRIPTION
**Pull request by binarymaster**
Backport ReactOS boot support commits, including fix in the GRUB wrapper code.

Third stage of three towards ReactOS boot support, fixes #17.

Important notes:
- `fsys_fat.c` and `fat.h` are imported as is from GRUB project
- I placed some `FIXME` comments regarding side problems, and I will convert them to issues later

### CODE REVIEW

File: include/memory_layout.h [hunk 1]
```diff
@@ -20,15 +20,19 @@
...
+
 #define MEMORYMANAGERSTART MAX_INITRD_END
 #define MEMORYMANAGEREND   0x039FFFFF
+#define MEMORYMANAGERSIZE  (MEMORYMANAGEREND - MEMORYMANAGERSTART)
 
 #define STACK_TOP 0x03C00000
 
-#define MAX_KERNEL_SIZE    (KERNEL_PM_CODE_END - KERNEL_PM_CODE)
-#define MEMORYMANAGERSIZE  (MEMORYMANAGEREND - MEMORYMANAGERSTART)
-#define MAX_INITRD_SIZE    (MAX_INITRD_END - INITRD_START)
```
Comment (JayFoxRox) [inline include/memory_layout.h]: Why was this moved?

File: boot/LoadReactOS.c [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+	"pushl %%eax \n"
+	"movl %0,%%eax \n"
+	"ret \n"
+	: : "i" (MULTIBOOT_BOOTLOADER_MAGIC), "b" (mbInfo), "a" (mbHeader->entry_addr));
+
+	// We are not longer here, we are already in freeldr, we never come back here
+
+	// See you again in ReactOS then
+	while (1);
+}
```
Comment (JayFoxRox) [inline boot/LoadReactOS.c]: This entire file (and commit) feels like a misnomer. This looks like it is about adding support for multiboot, not ReactOS. ReactOS support only appears to be the positive sideeffect and primary motivation of having multiboot. Or is multiboot exclusive to ReactOS (by design, or in this implementation)?

If this code is not exclusive to ReactOS, then it shouldn't refer to ReactOS explicitly.

File: fs/grub/disk_io.c [hunk 3]
```diff
@@ -128,30 +128,31 @@ int filemax;
...
 rawread (int drive, int sector, int byte_offset, int byte_len, char *buf)
 {
+	/* FIXME: Simplify this function, as BootIdeReadSector() supports offset and buffering */
 	sector+=byte_offset/512;
 	byte_offset%=512;
 	
 	while(byte_len) {
-		int nThisTime=512;
-	       	if(byte_len<512) nThisTime=byte_len;
+		int nThisTime = (byte_len < 512 ? byte_len : 512);
```
Comment (JayFoxRox) [inline fs/grub/disk_io.c]: nit: Parentheses not necessary. If you add them, I'd instead suggest `[...] = (condition) ? true_value : false_value;`

File: fs/grub/disk_io.c [hunk 4]
```diff
@@ -128,30 +128,31 @@ int filemax;
...
 			u8 ba[512];
-			if(BootIdeReadSector(drive, buf, sector, 0, 512)) {
-				bprintf("Unable to get first sector\n");
-                             	return 1;
+			if(BootIdeReadSector(drive, &ba, sector, 0, 512)) {
+				bprintf("Unable to read sector %d of IDE drive %d (512 bytes)\n", sector, drive);
+				return 1;
+			}
+			/* FIXME: Uncommenting next line breaks ReactOS boot from native FAT32 partition */
+			//nThisTime -= byte_offset;
```
Comment (JayFoxRox) [inline fs/grub/disk_io.c]: Why would this cause issues?
Not having it can cause out-of-bounds reads.

Sounds like an error elsewhere.

File: fs/grub/disk_io.c [hunk 5]
```diff
@@ -128,30 +128,31 @@ int filemax;
 int
 rawread (int drive, int sector, int byte_offset, int byte_len, char *buf)
 {
+	/* FIXME: Simplify this function, as BootIdeReadSector() supports offset and buffering */
```
Comment (JayFoxRox) [inline fs/grub/disk_io.c]: If this refactors this function anyway, why not fix this now?

File: include/memory_layout.h (hunk 1 above)
Comment (binarymaster) [reply include/memory_layout.h]: For better grouping of definitions.

Comment (JayFoxRox) [reply include/memory_layout.h]: Ah, I can see what you mean now - because kernel and initrd only affect Linux

File: boot/LoadReactOS.c (hunk 2 above)
Comment (binarymaster) [reply boot/LoadReactOS.c]: Yes it's not exclusive to ReactOS. But right now it's tied to the freeldr bootloader, and would require more refactoring to have universal multiboot support.

File: fs/grub/disk_io.c (hunk 4 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: Yes probably there is some other error elsewhere. See the notes from PR description.

File: fs/grub/disk_io.c (hunk 5 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: Same as above, see the notes from PR description.

File: fs/grub/disk_io.c (hunk 3 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: Improved.

File: boot/LoadReactOS.c [hunk 6]
```diff
@@ -0,0 +1,410 @@
...
+
+		if (CheckMultibootHeader(FREELDR_LOAD_AREA) == NULL) {
+			wait_ms(2000);
+			return false;
+		}
+		multiboot->pBuffer = FREELDR_LOAD_AREA;
+		multiboot->uBufferSize = dwSize;
+		/* Multiboot partition indices starting at 0, FreeLoader will increment 0xFE to 0xFF
+		 * and later it will interpret partition number 0xFF as CD drive */
+		multiboot->uBootDevice = GetMultibootBootDevice(0x80 + config->drive, 0xFE, 0xFF, 0xFF);
```
Comment (disean) [inline boot/LoadReactOS.c]: See comment.

Comment (binarymaster) [reply boot/LoadReactOS.c]: @disean fixed, thanks! Is the partition number ok, or it should be equal to 0?

Comment (disean) [reply boot/LoadReactOS.c]: That's correct now.

File: boot/LoadReactOS.c [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+		if (CheckMultibootHeader(FREELDR_LOAD_AREA) == NULL) {
+			wait_ms(2000);
+			return false;
+		}
+		multiboot->pBuffer = FREELDR_LOAD_AREA;
+		multiboot->uBufferSize = dwSize;
+		/* Multiboot partition indices starting at 0, FreeLoader will increment 0xFE to 0xFF
+		 * and later it will interpret partition number 0xFF as CD drive.
+		 * DL should be set to 0x80 for HDD, and 0xE0 for CD-ROM. */
```
Comment (JayFoxRox) [inline boot/LoadReactOS.c]: What is `DL` here? x86 register `DL`?
It's not obvious what this means - this is the C portion of the code, so it should explain stuff in C terms (probably speak about `drive` and `partition` parameters here - as the first part of the comment already does).

File: boot/LoadReactOS.c [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+		printk(" - %d bytes\n", fileinfo.fileRead);
+
+		if (CheckMultibootHeader(fileinfo.buffer) == NULL) {
+			wait_ms(2000);
+			return false;
+		}
+		multiboot->pBuffer = fileinfo.buffer;
+		multiboot->uBufferSize = fileinfo.fileRead;
+		/* Multiboot partition indices starting at 0, FATX (E:) is first one */
+		multiboot->uBootDevice = GetMultibootBootDevice(0x80 + partition->nDriveIndex, 0, 0xFF, 0xFF);
```
Comment (JayFoxRox) [inline boot/LoadReactOS.c]:
```c
// LoadReactOSNative
multiboot->uBootDevice = GetMultibootBootDevice(0x80 + config->drive         , config->partition, 0xFF, 0xFF);
// LoadReactOSFATX
multiboot->uBootDevice = GetMultibootBootDevice(0x80 + partition->nDriveIndex, 0                , 0xFF, 0xFF);
```

How would these 2 be distinguished by freeloader (or cromwell)?

If Native `config->partition` is 0 and `config->drive` is X, wouldn't that be equivalent to FATX where `partition->nDriveIndex` is X? Can this happen? If reactOS will boot *any* filesystem (by auto-detect) then we wouldn't need 2 functions in Cromwell probably (as it's misleading to explicitly pick FATX boot, but to end up with FAT32).

File: boot/LoadReactOS.c [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+		// force the HDD into a good mode 0x40 ==UDMA | 2 == UDMA2
+		nAta = 2; // best transfer mode without 80-pin cable
+	}
+	// nAta=1;
+	BootIdeSetTransferMode(0, 0x40 | nAta);
+	BootIdeSetTransferMode(1, 0x40 | nAta);
+
+	/* orange, people seem to like that colour */
+	setLED("oooo");
+
```
Comment (JayFoxRox) [inline boot/LoadReactOS.c]: Linux loader also sets framebuffer here, why isn't this done here?

File: boot/LoadReactOS.c [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+	/* disable interrupts */
+	asm volatile ("cli\n");
```
Comment (JayFoxRox) [inline boot/LoadReactOS.c]: Up until this point this, the function seems to be the same code as the Linux variant. We should eventually factor this out.

File: boot/LoadReactOS.c (hunk 2 above)
Comment (JayFoxRox) [reply boot/LoadReactOS.c]: Fair enough - I don't care enough for Cromwell to argue this.
Ideally we'd probably have this as `LoadMultiboot` or something, and then only some ReactOS specific functions for finding it, however, we might never need this and we can always just rename / refactor later.

File: boot/LoadReactOS.c (hunk 7 above)
Comment (binarymaster) [reply boot/LoadReactOS.c]: Yes, it's x86 register `DL` on the freeldr side, passed from Cromwell with multiboot, but here it's `drive` value. :ok_hand: :slightly_smiling_face:

Comment improved.

File: boot/LoadReactOS.c (hunk 8 above)
Comment (binarymaster) [reply boot/LoadReactOS.c]: Both Cromwell and FreeLoader supports only FATX data partition (E:), that's why the value is equal to 0 in `LoadReactOSFATX()`.

FreeLoader can autodetect partition scheme and the filesystem, this information is not passed via multiboot. I checked it boots correctly, either from MBR/FAT32 or Xbox(BRFR)/FATX.

Also FreeLoader will always boot from MBR HDD when possible, and only if MBR is not detected, it will try to probe Xbox partitions. So the collision here is not possible.

File: boot/LoadReactOS.c (hunk 9 above)
Comment (binarymaster) [reply boot/LoadReactOS.c]: On 64 MB Xbox the address value is not changed, even though it's reassigned. But it does change on 128 MB Xbox, and the `vesafb` Linux driver then performs video initialization again.

FreeLoader and ReactOS does not perform reinitialization, it only reuses existing framebuffer. It will work with 64 MB RAM, but won't work when the address change. There are also `FIXME` comments regarding framebuffer address explaining the problem.

File: boot/LoadReactOS.c (hunk 8 above)
Comment (JayFoxRox) [reply boot/LoadReactOS.c]: I feel like your explanation contradicts the code-comment which implies that "E:" is the first one, but then other partitions follow.

The function names also imply that this picks and depends on a partition scheme, whereas your reply suggests that this is merely an independent pre-test?

File: boot/LoadReactOS.c (hunk 9 above)
Comment (JayFoxRox) [reply boot/LoadReactOS.c]: I feel like this deserves a code comment. cromwell default was historically Linux, and people will probably compare code. There's also a risk that we'll factor out common code at some point, so we should note these differences and why they exist.

File: fs/grub/disk_io.c (hunk 4 above)
Comment (JayFoxRox) [reply fs/grub/disk_io.c]: This is really hard to accept. The Discord discussion is not linked, so this is without context and this touches code, has obvious issues, but is still supposed to be merged.

I'm willing to accept this under the following conditions:

- Only if this is read-only.
- This is using a moderately good IDE driver which stays read-only (won't corrupt peoples drives).
- If an issue with more context is created after merge.

File: fs/grub/disk_io.c (hunk 5 above)
Comment (JayFoxRox) [reply fs/grub/disk_io.c]: Same as https://github.com/XboxDev/cromwell/pull/27#discussion_r362119878

File: fs/grub/disk_io.c (hunk 4 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: - Yes, it's read-only.
- It seems like yes, driver is used in read-only mode.
- Issue will be created.

File: boot/LoadReactOS.c (hunk 8 above)
Comment (binarymaster) [reply boot/LoadReactOS.c]: Improved the comment to make things clear.

File: boot/LoadReactOS.c (hunk 9 above)
Comment (binarymaster) [reply boot/LoadReactOS.c]: Added a comment here.

File: fs/grub/disk_io.c (hunk 4 above)
Comment (mborgerson) [reply fs/grub/disk_io.c]: I'm against merging this particular piece of code as it is. The logic is just
flawed and having the commented-out line in the code causes more confusion.

Consider this case:

```c
byte_offset = 510
nThisTime = 10

BootIdeReadSector(drive, ba, sector, 0, 512);
nThisTime -= byte_offset; // <-- a
memcpy(out_buf, &ba[byte_offset], nThisTime)
```

If the line (a) is left in, now we get a negative value in `nThisTime`! If we
take the line out, then we get our 10 bytes, but we overread the block buffer
and start reading junk from the program stack. I guess we have gotten lucky in
testing with it.

Without getting into a total refactor of this code base, please consider this
alternative logic for now, which I think should meet the demands:

```c
// Possibly defined elsewhere
#define SECTOR_SIZE 512
#define MIN(x,y) ((x)<(y)?(x):(y))

int rawread(int drive, int sector, int byte_offset, int byte_len, char *buf)
{
	u8 sector_buf[SECTOR_SIZE];
	int status;
	int bytes_to_copy;

	// Advance sector counter by offset to get sub-sector-size offset
	sector += byte_offset/SECTOR_SIZE;
	byte_offset %= SECTOR_SIZE;

	// Handle first sector which may have an offset
	if (byte_offset) {
		// Determine max number of bytes to copy from this sector
		bytes_to_copy = MIN(byte_len, SECTOR_SIZE-byte_offset);

		// Read entire sector into temporary buffer
		status = BootIdeReadSector(drive, sector_buf, sector, 0, SECTOR_SIZE);
		if (status) goto error;

		// Copy sector data into dest buf
		memcpy(buf, &sector_buf[byte_offset], bytes_to_copy);
		buf += bytes_to_copy;
		byte_len -= bytes_to_copy;
		sector++;
	}

	// We are now sector-aligned. Read remainder directly into destination buffer
	while (byte_len > 0) {
		bytes_to_copy = MIN(byte_len, SECTOR_SIZE);
		status = BootIdeReadSector(drive, buf, sector, 0, bytes_to_copy);
		if (status) goto error;
		
		buf += bytes_to_copy;
		byte_len -= bytes_to_copy;
		sector++;
	}

	// I'm unsure about this return value, but have retained it from original
	// code
	return !(errnum);

error:
	bprintf("Unable to read sector %d of IDE drive %d (%d bytes)\n", sector, drive, bytes_to_copy);
	return 1;
}
```

File: boot/LoadReactOS.c [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+
+#include "boot.h"
+#include "BootFATX.h"
+#include "memory_layout.h"
+#include <shared.h>
+
+extern grub_error_t errnum;
+
+/* Length of area to be searched for multiboot header. */
+#define MULTIBOOT_SEARCHAREA_LEN        8192
```
Comment (mborgerson) [inline boot/LoadReactOS.c]: Ideally I'd like to see lots of this multiboot stuff refactored out, as it could be re-used for other multiboot-compliant binaries as well, but we can save that for a follow-up PR.

File: boot_xbe/xbe.S [hunk 12]
```diff
@@ -47,7 +47,7 @@ certificate:
 
   // title name (unicode string, 40 chars) *unimportant, but beautiful*
 unicodename:
-  .word 'L','i','n','u','x',0,0,0,0,0,0
+  .word 'X','r','o','m','w','e','l','l',0,0,0
```
Comment (mborgerson) [inline boot_xbe/xbe.S]: Would you mind changing this to **C**romwell instead of Xromwell? After previous discussion on Discord (I think) we agreed that this just adds confusion. Either Cromwell is loaded from an XBE or it's loaded on boot, either way it is still Cromwell.

File: boot/LoadReactOS.c (hunk 11 above)
Comment (JayFoxRox) [reply boot/LoadReactOS.c]: Also see this resolved discussion: https://github.com/XboxDev/cromwell/pull/27#discussion_r359862815.
I resolved that earlier multiboot discussion because it's out of scope for this PR and probably not necessary either. there's no immediate gain from generalizing multiboot due to lack of OS.

*(Meanwhile factoring out Linux / reactOS common code removes maintenance overhead; see https://github.com/XboxDev/cromwell/pull/27#discussion_r360167001 for that)*

File: boot_xbe/xbe.S (hunk 12 above)
Comment (JayFoxRox) [reply boot_xbe/xbe.S]: There are other instances of Xromwell in the repository (README at least), so I'd suggest to do this rebranding in a separate PR (before or after this one).

**Comment by mborgerson** [comment]
I have finished reviewing this PR. I've not tested it myself but I've skimmed over the boot stuff. If you would please have a look at the comments I left recently and fix up those items and test, I will merge this.

File: fs/grub/disk_io.c (hunk 4 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: @mborgerson your function worked perfectly, many thanks! I replaced it.

File: fs/grub/disk_io.c [hunk 13]
```diff
@@ -125,36 +125,30 @@ int filepos;
...
+		
+		buf += bytes_to_copy;
+		byte_len -= bytes_to_copy;
 		sector++;
+		byte_offset = 0;
 	}
-	return (!errnum);
+
+	// I'm unsure about this return value, but have retained it from original
+	// code
```
Comment (JayFoxRox) [inline fs/grub/disk_io.c]: This comment was probably never meant to make it into the code.

See https://stuff.mit.edu/afs/sipb/project/hurddev/tb/grub-0.5.92/stage2/disk_io.c line 96 and 189:

```c
  while (byte_len > 0 && !errnum)
```
```c
  return (!errnum);
```

So this functions assumes that no error was set beforehand and then keeps the state as is on success.

File: fs/grub/disk_io.c [hunk 14]
```diff
@@ -125,36 +125,30 @@ int filepos;
...
+	sector += byte_offset/SECTOR_SIZE;
+	byte_offset %= SECTOR_SIZE;
+
+	while (byte_len > 0) {
+		// Note: BootIdeReadSector will only read a single sector
+		int bytes_to_copy = MIN(byte_len, SECTOR_SIZE-byte_offset);
+		int status = BootIdeReadSector(drive, buf, sector, byte_offset, bytes_to_copy);
+		if (status != 0) {
+			bprintf("Unable to read sector %d of IDE drive %d (%d bytes)\n", sector, drive, bytes_to_copy);
+			return 1;
```
Comment (JayFoxRox) [inline fs/grub/disk_io.c]: This should ideally set an error message

See https://stuff.mit.edu/afs/sipb/project/hurddev/tb/grub-0.5.92/stage2/disk_io.c line 154:

```c
		    errnum = ERR_READ;
```

So this functions should set `errnum` in case of error, then `return 1`.

File: fs/grub/disk_io.c [hunk 15]
```diff
@@ -125,36 +125,30 @@ int filepos;
 int filemax;
 
 //Cromwell hooks
-int
-rawread (int drive, int sector, int byte_offset, int byte_len, char *buf)
+int rawread(int drive, int sector, int byte_offset, int byte_len, char *buf)
```
Comment (JayFoxRox) [inline fs/grub/disk_io.c]: The original code also seems to do sanity checking.

See https://stuff.mit.edu/afs/sipb/project/hurddev/tb/grub-0.5.92/stage2/disk_io.c line 93-94:

```c
  if (byte_len <= 0)
    return 1;
```

So for this case, the function should just `return 1` without setting `errnum`.
Our current function would return 0 in case of `byte_len == 0` instead.
Our current function would break in case of `byte_len < 0` instead.

File: fs/grub/disk_io.c [hunk 16]
```diff
@@ -125,36 +125,30 @@ int filepos;
...
-                      	byte_offset=0;
-             	} else {
-                     	if(BootIdeReadSector(drive, buf, sector, 0, nThisTime)) {
-                      		bprintf("Unable to get first sector\n");
-                             	return 1;
-                     	}
-                   	buf+=nThisTime;
-                   	byte_len-=nThisTime;
-              	}
+	// Advance sector counter by offset to get sub-sector-size offset
```
Comment (JayFoxRox) [inline fs/grub/disk_io.c]: nit: A comment that we assume that `byte_offset >= 0` would be good.

If this codebase uses `assert` (or similar), use that.

File: fs/grub/disk_io.c (hunk 5 above)
Comment (JayFoxRox) [reply fs/grub/disk_io.c]: Resolved in https://github.com/XboxDev/cromwell/pull/27#discussion_r362119878

File: boot/BootFromDevice.c [hunk 17]
```diff
@@ -271,6 +298,8 @@ int BootFromDevice(CONFIGENTRY *config) {
 	if (result) {
 		if (config->bootSystem == SYS_LINUX)
 			ExittoLinux(&config->opt.Linux);
+		if (config->bootSystem == SYS_REACTOS)
+			ExittoReactOS(&config->opt.Multiboot);
```
Comment (JayFoxRox) [inline boot/BootFromDevice.c]: nit: It's weird how we now have `bootSystem` as `REACTOS`, but use `opt.Multiboot` (`OPTMULTIBOOT`).

(Not sure if I had requested this, but this doesn't make much sense - it should be made consistent)

File: fs/grub/shared.h [hunk 18]
```diff
@@ -371,6 +371,9 @@ extern char *grub_scratch_mem;
 #define strcpy grub_strcpy
 #endif /* WITHOUT_LIBC_STUBS */
 
+#define SECTOR_SIZE 512
```
Comment (JayFoxRox) [inline fs/grub/shared.h]: This already exists in https://github.com/XboxDev/cromwell/blob/4b2db9caf5ef0e75a1a98a62b945ba9d8a3665e3/fs/grub/shared.h#L172

File: fs/grub/shared.h [hunk 19]
```diff
@@ -371,6 +371,9 @@ extern char *grub_scratch_mem;
...
+#define MIN(x,y) ((x)<(y)?(x):(y))
```
Comment (JayFoxRox) [inline fs/grub/shared.h]: nit: If this exists elsewhere, use it. Otherwise I'd suggest to move this into the file where it's used, or, to remove the macro altogether, as our use case can be dumped down into a single if statement along the lines of:

```
int bytes_available = SECTOR_SIZE - byte_offset;
int bytes_to_copy = byte_len;
if (bytes_to_copy > bytes_available) {
    bytes_to_copy = bytes_available;
}
```

(which is hopefully also self-documenting)

File: fs/grub/disk_io.c (hunk 13 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: Fixed.

File: fs/grub/disk_io.c (hunk 14 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: Fixed.

File: fs/grub/disk_io.c (hunk 15 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: Improved.

File: fs/grub/disk_io.c (hunk 16 above)
Comment (binarymaster) [reply fs/grub/disk_io.c]: Improved as well.

File: fs/grub/shared.h (hunk 18 above)
Comment (binarymaster) [reply fs/grub/shared.h]: Indeed, fixed.

File: fs/grub/shared.h (hunk 19 above)
Comment (binarymaster) [reply fs/grub/shared.h]: Moved macro to the source file.

File: boot/BootFromDevice.c (hunk 17 above)
Comment (binarymaster) [reply boot/BootFromDevice.c]: The structure itself doesn't contain any ReactOS specific values, that's why I called it like that. Also it gives an idea for future refactoring, as discussed in https://github.com/XboxDev/cromwell/pull/27#discussion_r363062315.

**Comment by mborgerson** [comment]
Thanks!
